prep/go/tutorial/chapter4/xkcd.go

```diff
@@ -0,0 +1,465 @@
+    return ids[:idsToReturn]
+}
+
+// Creates a simple case-insensitive inverted index of alphabetic terms
+func loadAndBuildIndex(comics *[]XKCDComic) ComicIndex {
+    indexData, err := os.ReadFile(COMIC_INDEX_STORAGE)
+    if err != nil {
+        if os.IsNotExist(err) {
+            indexData = []byte("{}")
+        } else {
+            log.Fatal(err)
+            os.Exit(1)
+        }
+    } 
+    index := newComicIndex()
+    err = json.Unmarshal(indexData, &index)
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+
+    for _, comic := range *comics {
+        if !comic.Indexed {
+            // index title and transcript
+            textToIndex := strings.Join([]string{comic.SafeTitle, comic.Title, comic.Transcript}, " ")
+            scanner := bufio.NewScanner(strings.NewReader(textToIndex))
+            scanner.Split(scanLetters)
+            for scanner.Scan() {
+                term := strings.ToLower(scanner.Text())
+                // Skip words like "a" and "to"
+                if len(term) < 3 {
+                    continue
+                }
+                addComicIndex(index, term, comic.Id)
+            }
+
+            // TODO flush comics back to disk with Index = True so it
+            // doesn't get indexed every run
+            // Trying to do it as everything is written now introduces a bug
+            // that isn't worth fixing until the performance is desired (profiling exercise? :))
+            comic.Indexed = true 
+        }
+    }
+    flushIndexToDisk(&index)
+    return index
+}
+
+// TODO ChatGPT'd this. Need to understand writing custom scanners outside this exercise
+func scanLetters(data []byte, atEOF bool) (advance int, token []byte, err error) {
+    start := 0
+    for ; start < len(data); start++ {
+        if unicode.IsLetter(rune(data[start])) {
+            break
+        }
+    }
+
+    for i := start; i < len(data); i++ {
+        if !unicode.IsLetter(rune(data[i])) {
+            return i + 1, data[start:i], nil
+        }
+    }
+
+    
+    if !atEOF && unicode.IsLetter(rune(data[len(data)-1])) {
+        return start, nil, nil
+    }
+
+    return len(data) + 1, data[start:], nil
+}
+
+func flushIndexToDisk(index *ComicIndex) {
+    data, err := json.MarshalIndent(index, "", " ")
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    err = ioutil.WriteFile(COMIC_INDEX_STORAGE, data, 0644)
+    if err != nil {
+        log.Fatal(err)
+    }
+}
+
+
+func loadComics(comics *[]XKCDComic) {
+    var wg sync.WaitGroup
+    // Comics start at 1
+    i := 1
+    finished := false
+    for {
+        // Very funny xkcd.. Very funny..
+        if i == 404 {
+            i++
+        }
+
+        wg.Add(1)
+        // id is incrementing and passed by value so it is safe
+        // each routine will only access its index in the array
+        go tryLoadFromDiskFallbackToInternet(i, &(*comics)[i], &wg, &finished)
+
+        // Run lots of go routines for the first 2500
```

Comment:
An alternative here is to enforce that "only N goroutines can be concurrently trying to download at a time". One way to achieve this is by creating a buffered channel with buffer size N, and have each goroutine do something like this:

1. send on the channel
2. perform the download
3. receive from the channel

Step 1 will block if N downloads are currently active

Reply:
Channels may be the way to do this. Jumped ahead here and didn't give them a good read. I think that'd make the implementation more clear.